pkg/controller/postgres/postgres_controller.go

```diff
@@ -140,221 +158,507 @@ func (r *ReconcilePostgres) Reconcile(request reconcile.Request) (reconcile.Resu
+	if statefulSet.Spec.Replicas != nil {
+		intendentReplicas = *statefulSet.Spec.Replicas
+	}
+
+	if statefulSet.Status.ReadyReplicas == intendentReplicas {
+		postgres.Status.Active = true
+	}
+
+	return reconcile.Result{}, r.client.Status().Update(context.Background(), postgres)
 }
 
-func (r *ReconcilePostgres) updateStatus(
-	postgres *contrail.Postgres, pod *core.Pod, cip string,
-) error {
-	err := r.client.Get(context.Background(), types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, pod)
+func (r *ReconcilePostgres) ensureLabelExists(p *contrail.Postgres) error {
+	if len(p.Labels) != 0 {
+		return nil
+	}
+
+	p.Labels = contraillabel.New(contrail.PostgresInstanceType, p.Name)
+	return r.client.Update(context.Background(), p)
+}
+
+func (r *ReconcilePostgres) ensureCertificatesExist(postgres *contrail.Postgres, pods *core.PodList, serviceIP string) error {
+	hostNetwork := true
+	if postgres.Spec.CommonConfiguration.HostNetwork != nil {
+		hostNetwork = *postgres.Spec.CommonConfiguration.HostNetwork
+	}
+	return certificates.NewCertificateWithServiceIP(r.client, r.scheme, postgres, pods, serviceIP, "postgres", hostNetwork).EnsureExistsAndIsSigned()
+}
+
+func (r *ReconcilePostgres) listPostgresPods(postgres *contrail.Postgres) (*core.PodList, error) {
+	pods := &core.PodList{}
+	labelSelector := labels.SelectorFromSet(postgres.Labels)
+	listOpts := client.ListOptions{LabelSelector: labelSelector}
+	if err := r.client.List(context.TODO(), pods, &listOpts); err != nil {
+		return &core.PodList{}, err
+	}
+	return pods, nil
+}
+
+func (r *ReconcilePostgres) ensureServicesExist(postgres *contrail.Postgres) (*core.Service, error) {
+	service := &core.Service{
+		ObjectMeta: meta.ObjectMeta{
+			Name:      postgres.Name,
+			Namespace: postgres.Namespace,
+		},
+	}
+	_, err := controllerutil.CreateOrUpdate(context.Background(), r.client, service, func() error {
+		service.ObjectMeta.Labels = postgres.Labels
+		service.Spec.Type = core.ServiceTypeClusterIP
+		nodePort := int32(0)
+		listenPort := int32(postgres.Spec.ServiceConfiguration.ListenPort)
+		for i, p := range service.Spec.Ports {
+			if p.Port == listenPort {
+				nodePort = service.Spec.Ports[i].NodePort
+			}
+		}
+		service.Spec.Ports = []core.ServicePort{
+			{Port: listenPort, Protocol: "TCP", NodePort: nodePort},
+		}
+		return controllerutil.SetControllerReference(postgres, service, r.scheme)
+	})
+
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	if len(pod.Status.ContainerStatuses) != 0 {
-		postgres.Status.Active = pod.Status.ContainerStatuses[0].Ready
-		postgres.Status.Endpoint = cip
-	} else {
-		postgres.Status.Active = false
+	serviceRepl := &core.Service{
+		ObjectMeta: meta.ObjectMeta{
+			Name:      postgres.Name + "-replica",
+			Namespace: postgres.Namespace,
+		},
 	}
-	return r.client.Status().Update(context.Background(), postgres)
+
+	_, err = controllerutil.CreateOrUpdate(context.Background(), r.client, serviceRepl, func() error {
```

Comment:
Please use https://github.com/Juniper/contrail-operator/blob/master/pkg/k8s/service.go

Reply:
You can modify it a bit if you need that

Reply:
done